lib/events/azsessions/azsessions.go

```diff
@@ -0,0 +1,549 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/streaming"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/session"
+)
+
+const (
+	sessionContainerName    = "session"
+	inprogressContainerName = "inprogress"
+
+	// clientIDFragParam is the parameter in the fragment that specifies the optional client ID.
+	clientIDFragParam = "azure_client_id"
+)
+
+// sessionName returns the name of the blob that contains the recording for a
+// given session.
+func sessionName(sid session.ID) string {
+	return sid.String()
+}
+
+// uploadMarkerPrefix is the prefix of the names of the upload marker blobs.
+// Listing blobs with this prefix will return an empty blob for each upload.
+const uploadMarkerPrefix = "upload/"
+
+// uploadMarkerName returns the blob name for the marker for a given upload.
+func uploadMarkerName(upload events.StreamUpload) string {
+	return fmt.Sprintf("%v%v/%v", uploadMarkerPrefix, upload.SessionID, upload.ID)
+}
+
+// partPrefix returns the prefix for the upload part blobs for a given upload.
+// Listing blobs with this prefix will return all the parts that currently make
+// up the upload.
+func partPrefix(upload events.StreamUpload) string {
+	return fmt.Sprintf("part/%v/%v/", upload.SessionID, upload.ID)
+}
+
+// partName returns the name of the blob for a specific part in an upload.
+func partName(upload events.StreamUpload, partNumber int64) string {
+	return fmt.Sprintf("%v%v", partPrefix(upload), partNumber)
+}
+
+// field names used for logging
+const (
+	fieldSessionID  = "session_id"
+	fieldUploadID   = "upload_id"
+	fieldPartNumber = "part"
+	fieldPartCount  = "parts"
+)
+
+// Config is a struct of parameters to define the behavior of Handler.
+type Config struct {
+	// ServiceURL is the URL for the storage account to use.
+	ServiceURL url.URL
+
+	// ClientID, when set, defines the managed identity's client ID to use for
+	// authentication.
+	ClientID string
+
+	// Log is the logger to use. If unset, it will default to the global logger
+	// with a component of "azblob".
+	Log logrus.FieldLogger
+}
+
+// SetFromURL sets values in Config based on the passed in URL: the fragment of
+// the URL is parsed as if it was made out of query parameters, which define
+// options for ourselves, and then the remainder of the URL is set as the
+// service URL.
+func (c *Config) SetFromURL(u *url.URL) error {
+	if u == nil {
+		return nil
+	}
+
+	c.ServiceURL = *u
+
+	switch c.ServiceURL.Scheme {
+	case teleport.SchemeAZBlob:
+		c.ServiceURL.Scheme = "https"
+	case teleport.SchemeAZBlobHTTP:
+		c.ServiceURL.Scheme = "http"
+	}
```

Comment:
Should this switch have a `default` clause that returns an error is there's an unsupported scheme?

Reply:
This is only accessible through `audit_sessions_uri` where the possible URL schemas are those two, or via custom code where the caller might want to do something weird. In practice getting an error here or at the first use of an unsupported scheme doesn't change things, so I think we can leave it like this.